### ISSUE TITLE
Feature request- Select whole words and other selection options

### ISSUE THREAD

**Issue by Yekutiel** [yurikhan/kitty_grab#7]
@yurikhan 

kitty-grab works great! Thank you! I found your documentation easy to follow. kitty-grab should be included in kitty itself. I detest using my mouse to copy text.

I'd like to use kitty-grab just as I use following standard text selection features on my Linux machine when I use Google Docs, Mousepad, SpaceFM, LibreWriter, etcetera...

1. `Shift+Ctrl+Right` to select one word to the right.
2. `Shift+Ctrl+Left` to select one word to the left.
3. `Shift+End` to select from the current cursor position to the end of the line.
4. `Shift+Home` to select from the current cursor position to the beginning of the line.
5. `Shift+Ctrl+End` to select from the current cursor position to the end of the "document."
6. `Shift+Ctrl+Home` to select from the current cursor position to the beginning of the "document."

Thanks,

Yekutiel

**Comment by yurikhan** [comment]
I’m missing some of these too, but there are complications.

* Word motion: Should probably honor Kitty’s `select_by_word_characters` configuration parameter.
* What do we consider the beginning of line? Position 0 or first non-whitespace character?
* What do we consider the end of line? Last non-whitespace character? What if there is a box around the whole screen, as in e.g. Midnight Commander, does that mean the end of line is at the right edge of the window?
* What is the end of the “document”? End of the last line in the scrollback? End of the last non-blank line in the scrollback? One newline beyond the end of the last non-blank line in the scrollback?
* What is the beginning of the “document”? Start of the first line in the scrollback, all the way back, however many lines are retained (`scrollback_lines` or `scrollback_pager_history_size`)? Is that useful?

**Comment by Yekutiel** [comment]
@yurikhan 

Thanks for your quick, helpful, and detailed response.

> Word motion: Should probably honor Kitty’s select_by_word_characters configuration parameter.

That sounds fine. See, I very rarely choose individual characters. The smallest unit I normally choose is a word, not a character.



> What do we consider the beginning of line? Position 0 or first non-whitespace character?

That's a very good question which I had not considered. Let's consider the following line of text from a kitty session...
![2019-11-12_04-34](https://user-images.githubusercontent.com/16394972/68672122-cff2e980-0505-11ea-9377-1034a5ab6c6d.png)

Ideally I'd like to be able to choose either of the following with the indicated keybindings...

**Alt+Shift+Home** to choose...
`[y@y-pc ~]$ This is a sample line of text.`

**Shift+Home** to choose...
`This is a sample line of text.`


> What do we consider the end of line? Last non-whitespace character? 

I think we should consider the last non-whitespace character to be the end of the line.


> What if there is a box around the whole screen, as in e.g. Midnight Commander, does that mean the end of line is at the right edge of the window?

I think I installed Midnight Commander but never used it because I detest looking at console applications (unless I have to) and looking at two panels disorients me. If I had to use a console application as a file manager I'd probably use Ranger. But I don't need to use a a console application as a file manager because I use SpaceFM which works great for me (when combined with AutoKey and wmctrl). Yes, using wmctrl requires me to use a little Python, but I use Google to find existing Python code other people have written. I barely understand Python.


> box around the whole screen

I'm not sure what you mean by a "box around the whole screen."

Please let me explain. I'm not an engineer. In other words, I'm not a developer. I actually like GUIs for choosing commands I use infrequently. For example, Vim drives me crazy. Nevertheless I end up using myriad keyboard shortcuts via AutoKey, Xbindkeys, Xmodmap, ClipIt, and SpaceFM. More to the point of our discussion herein: for selecting text I rarely use a mouse. For selecting text (which I do very frequently) I use combinations of Right, Left, Up, Down, Ctrl, Shift, PageUp, PageDown, Home, and End. I press those eight keys on my keyboard frequently throughout the day. I write a lot and constantly need to edit what I'm writing. I await the day when Google Docs' voice recognition of voice commands allows me to select text without relying on those eight keys. But that day has not yet arrived.

Sooooo I repeat....

I'm not sure what you mean by a "box around the whole screen."


> What is the end of the “document”? 

Again, to me, a person who spends a lot of time using Google Docs (I like Google Docs very much) the end of the document, is, well, it's the end of the document. I hope you understand what I mean. Again, I'm not a software developer. Therefore, for example, I neither use Vim nor Sublime. I am blissfully completely unfamiliar with Vim macros and plugins, and hope to remain completely unfamiliar with them for the rest of my life. 


> End of the last line in the scrollback? 

*Scrollback** is like Chinese to me. Again, I'm not a developer. I don't know what scrollback is. (I don't want to know what scrollback is).


> End of the last non-blank line in the scrollback? 

Again, that's like Chinese to me.


> One newline beyond the end of the last non-blank line in the scrollback?

Once again, that's like Chinese to me.


> What is the beginning of the “document”? 

That's easy. When I open Google Docs (or Kitty) **the beginning of the “document”** is the first letter of the first word I type. For example, in the following case...

![2019-11-12_04-34](https://user-images.githubusercontent.com/16394972/68673376-7213d100-0508-11ea-80bb-134540ad65d8.png)

the letter "T" in the word "This" would be **the beginning of the “document”** 


> Start of the first line in the scrollback, all the way back, however many lines are retained (scrollback_lines or scrollback_pager_history_size)? Is that useful?

Again, that's like Chinese to me.

In closing, why not follow the "kitty model."

I simply went to...

[Configuring kitty](https://sw.kovidgoyal.net/kitty/conf.html#configuring-kitty) and created...

[kitty.conf](https://pastebin.com/gTn7edmJ)

I have similarly customized config files for LibreOffice Writer, mpv and MPlayer. By the way, the news of MPlayer's death is greatly exaggerated (I tip my hat, [hat tip](https://en.wikipedia.org/wiki/Hat_tip), to whoever misattributed a [similar quotation](http://www.thisdayinquotes.com/2010/06/reports-of-my-death-are-greatly.html) to Mark Twain). MPlayer plays videos much smoother on my machine than mpv, but mpv offers me greater customization options which I find useful when watching video tutorials that I like to navigate through quickly.

Thanks,

Yekutiel

**Comment by yurikhan** [comment]
Okay, see, the principal reason I wrote this grabber is so that I, personally, can use it. So, the first few things implemented are the things I absolutely cannot live without (because, same as you, I do not want to use the mouse to select text in terminal). Then, I make it available to others, as they say, “in the hope that it will be useful, but without any warranty”.

> Alt+Shift+Home to choose...
> `[y@y-pc ~]$ This is a sample line of text.`
>
> Shift+Home to choose...
> `This is a sample line of text.`

Suddenly, a simple exercise becomes a research problem. All the grabber sees is a grid of characters. It does not know which of them you typed and which of them constitute the shell prompt.

> I detest looking at console applications (unless I have to) and looking at two panels disorients me.

Here, then, have a look at my typical terminal.

![image](https://user-images.githubusercontent.com/238691/68676599-f9a32380-058c-11ea-98ff-2569bdcd85f4.png)

This is Midnight Commander, with two file panels and a command line. (My choice of Midnight Commander as a file manager is defined by the presence of the command line, available in zero key presses — I can start typing and it goes to the command line. I have tried Ranger; there, in order to run a command, I have to press <kbd>!</kbd> before I can start typing a command.)

Anyway, what makes MC relevant in the context of this issue, is the border box around the panels: Because it’s always there, it becomes the last non-whitespace character of every line.

Observe also that MC has rows and columns. It is understandable that a user could want <kbd>End</kbd> to go to the last non-whitespace character in the current cell; but a simple grabber will not know where each column starts and ends.

> I'm not a developer. I don't know what scrollback is. (I don't want to know what scrollback is).

Oh. Well, here it goes anyway.

Your screenshot here shows a single line of text. Maybe that is a typical terminal session for you. For me, it’s not.

I *am* a developer. On a typical work day, I may run some fifty commands, each producing, let’s say for the sake of exposition, a few thousand lines of output. My terminal is only about 70 lines high. This means not all of the output fits in the window. So, when several things go wrong, I have to scroll the window back, in order to see the first one (which might have caused other errors). How far back I can go is called the scrollback buffer size.

What I’m saying is, conceptually, there *is* no start of document. There is an end; it’s what was written last. Sometimes there is a beginning — when I have just rebooted my computer (which happens roughly once a month). Otherwise, the real beginning has been scrolled off the scrollback buffer, and lost in time. All I have is some arbitrary point a few days ago.

Do you never have to go back to see the results of your last command?

> In closing, why not follow the "kitty model."

You mean, expose a programming interface so that everyone could define their own ideas of “start of document” and “end of line”? :trollface:


More to the point: I am not asking the questions above to demonstrate superiority. You are asking for a feature that I also find desirable, but I’m asking you to make some effort to produce a more precise description of your idea how it should work, considering the limitations of the data that the grabber sees (a simple rectangular grid of characters).

**Comment by Yekutiel** [comment]
@yurikhan 

> Here, then, have a look at my typical terminal.

I would lose my mind looking at your typical terminal. Why does every engineer seem to choose dark themes? Is it easier to look at for you guys? Dark themes look ominous to me. And two panels? That's information overload for me. And small fonts? Hard to look at and, again, too much information for me. And the general techie-looking terminal set up you have in Midnight Commander? That would drive me crazy.

> Suddenly, a simple exercise becomes a research problem. 

I see.


> All the grabber sees is a grid of characters. It does not know which of them you typed and which of them constitute the shell prompt. 

I presume that currently that assertion is correct. Perhaps the concepts I indicated below might enable you to transform you "the current blind kitty-grab" into "a kitty-grab that could see" without having to deal with a research problem.

I now realize that for you, in terms of copying and pasting, generally the beginning of a terminal session is meaningless and not at all analogous to my concept of an entire document because, whereas, for me a document is a discrete and clearly defined entity, for you a terminal session is tantamount to a series of concatenated information which a computer program could not meaningfully parse unless it had the powers of clairvoyance (could read your mind). 

By the way, I am not criticizing you in any manner whatsoever but rather I am merely observing. It seems that you, and I suppose most other software developers, use a terminal much like an army drill instructor uses a megaphone: as a way to issue commands.

> You mean, expose a programming interface so that everyone could define their own ideas of “start of document” and “end of line”?

No and no.

When I use a terminal I constantly close and clear my sessions because I like looking at blank screens. Although I don't save them, conceptually I think of each terminal session as akin to a document. The idea of creating a terminal session with copious amounts of information is anathema to me.

I like Google Docs for many reasons. One of my favorite features of Google Docs is the ability to create hyperlinks between Google Docs. I like to create short Google Docs. Many of my Google Docs pages look sort of like Wikipedia articles; that is, many of my Google Docs pages contains many links but not too much text. I'd rather create 25 or 30 brief Google Docs than one or two large Google Docs. Typing up this response to you was frustrating and difficult for me. I would have preferred to have simply created 10 or 20 Google Docs instead.

As another aside, I've been astounded to see computer programs with hundreds or even thousands of lines. I would be overwhelmed trying to deal with information presented in that manner. I'm able to much more readily link many discrete components together (like a typical Wikipedia article). Normally, once I create and understand small component I give it a name and the put the details about it elsewhere. Henceforth, refer to it by its name. If I need to review or revise the details I look at them in the small, discrete, straightforward, document I created on the subject.

And yes, as you supposed, my lines are usually very short (perhaps an average of not more than 100 to 200 characters). I'm not a developer. I don't do much in a terminal. Although I am going to need to set up a couple of servers and intend to create some Ansible playbooks.

In my mind, the “start of document” was not many thousands of lines back from, say, 10 or 20 days ago. To me, the “start of document” was, say, 20 or 30 lines back from, say, 30 to 45 minutes ago.

As for “end of line”, I'm hoping you might enable kitty-grab to see it.

Be that as it may, at least for now, I think it might be best to ignore the concept of "selecting an entire document" and instead focus on the concept of what you described as "scrolling the window back."

> Anyway, what makes MC relevant in the context of this issue, is the border box around the panels: Because it’s always there, it becomes the last non-whitespace character of every line.

Perhaps in `grab.conf` you might allow users to enter a value of **0**, or **1**, or **2**, and so on, to eliminate none, one, or two, and so on, of the non-whitespace character(s) at the end of every line.

I suppose that scrolling the window back a certain number of lines (for example, 50 lines or perhaps 100 lines) from the cursor's current location when a user were to press, for example, `Ctrl+PageUp` would probably be a requirement you could implement without too much trouble because I presume you would implement it using kitty-grab's current commands: `scroll up`, `scroll down`, `select stream left`, and `select stream right`. Right?

Here's something you might also consider...

*** **BEGINNING OF DEFINITIONS** ***

`The Cursor's Current Location`= the place in the line where the cursor is currently located (which, for example, would be the location where the cursor would be flashing in case, for example, a properly constructed `kitty.conf` file  were to contain `cursor_blink_interval -1` and `cursor_stop_blinking_after 15.0` in cases when not quite 15 seconds had yet to elapse from the time when the cursor had been moved to its current location. (By the way, as an example of how I like to work, I would normally have created a Google Doc called **The Cursor's Current Location** and hyperlinked to it from `The Cursor's Current Location` above).

`The Cursor's Current Line`= the line wherein the cursor is currently located.

`The Beginning of The Cursor's Current Line`= the location where the cursor would normally end up in cases when a user were to have pressed Ctrl+a or the Home key.

`The End of The Cursor's Current Line`= the location where the cursor would normally end up in cases when a user were to have pressed Ctrl+e or the End key.

*** **END OF DEFINITIONS** ***


Perhaps kitty-grab might  programmatically enter Ctrl+a and Ctrl+e prior to entering the "kity-grab screen" (which is what I think of as the screen kitty-grab creates) so as to essentially orient itself (anthropomorphizing intentional) within `The Cursor's Current Line`.  If such an orientation were accomplished then I presume a user could cause kitty-grab to choose...

1. `The Cursor's Current Line`, and/or
2. That portion of `The Cursor's Current Line` bounded at one extreme by `The Cursor's Current Location` and bounded at the other extreme by `The Beginning of The Cursor's Current Line` and/or
3. That portion of `The Cursor's Current Line` bounded at one extreme by `The Cursor's Current Location` and bounded at the other extreme by `The End of The Cursor's Current Line`

By applying this orientation technique I suppose a user could cause kitty-grab to choose entire lines when the user were to chose to scrollback. In other words, this might prevent kitty-grab from "blindly" only grabbing part of the first line and part of the the last line in a scrollback.

In addition to 'Spacial Selections' (which is what I discussed immediately above), I wonder if it might be possible to allow users to choose 'Temporal Selections' such as "the last 20 minutes" or "everything today". If  'Temporal Selections' were created then they I suppose they might be able to be combined with 'Spacial Selections' so that users could choose, for example, "scrollback not more than 200 lines beginning 40 minutes ago and ending up to, but not later, than 120 minutes ago."

I look forward to your response.

Thanks, 

Yekutiel

**Comment by yurikhan** [comment]
You use your computer very differently from the way I do.

> Perhaps kitty-grab might programmatically enter Ctrl+a and Ctrl+e prior to entering the "kity-grab screen" (which is what I think of as the screen kitty-grab creates) so as to essentially orient itself (anthropomorphizing intentional) within The Cursor's Current Line.

This would be possible if all interactive programs that you ran in a terminal adhered to the convention that <kbd>Ctrl+a</kbd> and <kbd>Ctrl+e</kbd> move the cursor to the start and end of the current line, respectively. This is of course not true; see, for a counterexample, Vim (which drives you crazy, which means you use it).

Also, even if you did that, that would limit you to selecting and copying your own input, whereas it is equally important to be able to copy output produced by programs.

**Comment by Yekutiel** [comment]
@yurikhan 

Thanks for your quick and helpful response.

As it is, I like kitty-grab very much. I hope you realize that I am trying to figure out how to make it quicker and easier to select text with kitty-grab.


> This would be possible if all interactive programs that you ran in a terminal adhered to the convention that Ctrl+a and Ctrl+e move the cursor to the start and end of the current line, respectively. This is of course not true; see, for a counterexample, Vim...

I see.


> ... (which drives you crazy, which means you use it).

I have used Vim but I very rarely use it. Instead of using Vim, when  I have the need (which is rare), I decided to download and save files from a server I am using to my local machine, edit them with Mousepad (or Leafpad), and then copy those files up to the same server I am using. I'm comfortable using Mousepad (or Leafpad) and SpaceFM. 

Trying to use Vim (or Nano) or other terminal based text editors along with a terminal to edit, delete, move, etcetera those files was very time-consuming for me because I rarely need to edit, delete, move, etcetera files files on a server.

Ok. Here's another thought...

How about being able to select whole words with kitty-grab? For example in many ordinary text editors (as well as this text box into which I am currently entering text) by holding down, "Ctrl+Shift+Right Arrow" I can essentially select one word at a time.

In other words, I am thinking that option would enable kitty-grab to select whole words instead the current option of selecting one character at a time.

Thanks,

Yekutiel

**Comment by yurikhan** [comment]
Selecting by words is a well-defined task and a useful feature. I will get around to it eventually.

**Comment by Yekutiel** [comment]
@yurikhan 

Thanks. I am looking forward to using that feature.

Bye for now,

Yekutiel

**Comment by Yekutiel** [comment]
>> Selecting by words is a well-defined task and a useful feature. I will get around to it eventually.

Why did you close this task? Apparently you have not added this feature yet.

**Comment by yurikhan** [comment]
I have.

**Comment by Yekutiel** [comment]
Ah. Ok. Thank you!
